### PR TITLE
GZIP content length mismatch fix

### PR DESCRIPTION
**Pull request by hiranp**
This addresses concerns raised by the following discussions:
 - [Gzip Middleware content-length is incorrect #803](https://github.com/encode/starlette/issues/803)
 - [GZip Middleware Throws "Response content longer than Content-Length" when given a 304 empty-body response #4050](https://github.com/tiangolo/fastapi/issues/4050)
 - [Gzip middleware randomly raise RuntimeError but does not affect use , what are the possible causes?](https://github.com/tiangolo/fastapi/issues/2818)

We encountered this problem when developing our internal application, when we implemented both Brotli and GZIP compression. We tested the following fix using several use cases.

The solution tries to address the problem using two(2) approaches:
- Setting the respond body if NULL bytes are detected
- Not writing the message body when its empty or contains NULL byte char

- [x] Initially raised as discussion #803

### CODE REVIEW

**Comment by Kludex** [comment]
Is this still an issue on https://github.com/tiangolo/fastapi/pull/4488 ?

**Comment by hiranp** [comment]
Yeah, does seem that the issue is not fully resolved in the 0.19.0 branch. 

You can try to reproduce the problem using the sample test code from: https://github.com/tiangolo/fastapi/issues/4050

Installed 0.19.0 Branch:
`pip install git+https: //github.com/Kludex/starlette.git@release/0.19.0
Collecting git+https: //github.com/Kludex/starlette.git@release/0.19.0
  Cloning https: //github.com/Kludex/starlette.git (to revision release/0.19.0) to /private/var/folders/x9/w4nnqprj1253mcyc4z_j_nx00000gp/T/pip-req-build-yllbbcv8
  Running command git clone --filter=blob:none --quiet https: //github.com/Kludex/starlette.git /private/var/folders/x9/w4nnqprj1253mcyc4z_j_nx00000gp/T/pip-req-build-yllbbcv8
  Running command git checkout -b release/0.19.0 --track origin/release/0.19.0
  Switched to a new branch 'release/0.19.0'
  branch 'release/0.19.0' set up to track 'origin/release/0.19.0'.
  Resolved https: //github.com/Kludex/starlette.git to commit 01e183cf348c7cbcb9fd77d41dcd657df4d616a0
  Preparing metadata (setup.py) ... done
Requirement already satisfied: anyio<5,>=3.4.0 in ./.venv/lib/python3.10/site-packages (from starlette==0.19.0) (3.5.0)
Requirement already satisfied: sniffio>=1.1 in ./.venv/lib/python3.10/site-packages (from anyio<5,>=3.4.0->starlette==0.19.0) (1.2.0)
Requirement already satisfied: idna>=2.8 in ./.venv/lib/python3.10/site-packages (from anyio<5,>=3.4.0->starlette==0.19.0) (3.3)
Building wheels for collected packages: starlette
  Building wheel for starlette (setup.py) ... done
  Created wheel for starlette: filename=starlette-0.19.0-py3-none-any.whl size=60789 sha256=20118f679cc0c053119465ce2b2cbd73d2b529ad55286e2b5c6db74c3d2feba2
  Stored in directory: /private/var/folders/x9/w4nnqprj1253mcyc4z_j_nx00000gp/T/pip-ephem-wheel-cache-6zyvvodd/wheels/90/a4/2e/2051f1b9eece9835113722a1e6da7af5b97e2bc7845e7eaa23
Successfully built starlette
Installing collected packages: starlette
  Attempting uninstall: starlette
    Found existing installation: starlette 0.17.1
    Uninstalling starlette-0.17.1:
      Successfully uninstalled starlette-0.17.1`

It seems that responses.py[74] is not handling the NoneType corner case, when a null is received in the byte stream. 
```
        if (
            body is not None
            and populate_content_length
            and not (self.status_code < 200 or self.status_code in (204, 304))
        ):
```

Stack Trace:
```
[04/08/22 10:55:02] ERROR    Exception in ASGI application                                                                                                                              
                             Traceback (most recent call last):                                                                                                                             
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/anyio/streams/memory.py", line 81, in receive                                       
                                 return self.receive_nowait()                                                                                                                               
                                        |    -> <function MemoryObjectReceiveStream.receive_nowait at 0x106367d00>                                                                          
                                        -> MemoryObjectReceiveStream(_state=MemoryObjectStreamState(max_buffer_size=0, buffer=deque([]), open_send_channels=0, open_rece...                 
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/anyio/streams/memory.py", line 76, in receive_nowait                                
                                 raise WouldBlock                                                                                                                                           
                                       -> <class 'anyio.WouldBlock'>                                                                                                                        
                                                                                                                                                                                            
                             anyio.WouldBlock                                                                                                                                               
                                                                                                                                                                                            
                                                                                                                                                                                            
                             During handling of the above exception, another exception occurred:                                                                                            
                                                                                                                                                                                            
                                                                                                                                                                                            
                             Traceback (most recent call last):                                                                                                                             
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 41, in call_next                                
                                 message = await recv_stream.receive()                                                                                                                      
                                                 |           -> <function MemoryObjectReceiveStream.receive at 0x106367d90>                                                                 
                                                 -> MemoryObjectReceiveStream(_state=MemoryObjectStreamState(max_buffer_size=0, buffer=deque([]), open_send_channels=0, open_rece...        
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/anyio/streams/memory.py", line 101, in receive                                      
                                 raise EndOfStream                                                                                                                                          
                                       -> <class 'anyio.EndOfStream'>                                                                                                                       
                                                                                                                                                                                            
                             anyio.EndOfStream                                                                                                                                              
                                                                                                                                                                                            
                                                                                                                                                                                            
                             During handling of the above exception, another exception occurred:                                                                                            
                                                                                                                                                                                            
                                                                                                                                                                                            
                                                                                                                                                                                            
                             Traceback (most recent call last):                                                                                                                             
                                                                                                                                                                                            
                               File "<string>", line 1, in <module>                                                                                                                         
                                                                                                                                                                                            
                               File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main       
                                 exitcode = _main(fd, parent_sentinel)                                                                                                                      
                                            |     |   -> 3                                                                                                                                  
                                            |     -> 17                                                                                                                                     
                                            -> <function _main at 0x105743e20>                                                                                                              
                                                                                                                                                                                            
                               File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/spawn.py", line 129, in _main            
                                 return self._bootstrap(parent_sentinel)                                                                                                                    
                                        |    |          -> 3                                                                                                                                
                                        |    -> <function BaseProcess._bootstrap at 0x10530e5f0>                                                                                            
                                        -> <SpawnProcess name='SpawnProcess-1' parent=86862 started>                                                                                        
                                                                                                                                                                                            
                               File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/process.py", line 315, in _bootstrap     
                                 self.run()                                                                                                                                                 
                                 |    -> <function BaseProcess.run at 0x10530dc60>                                                                                                          
                                 -> <SpawnProcess name='SpawnProcess-1' parent=86862 started>                                                                                               
                                                                                                                                                                                            
                               File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/process.py", line 108, in run            
                                 self._target(*self._args, **self._kwargs)                                                                                                                  
                                 |    |        |    |        |    -> {'config': <uvicorn.config.Config object at 0x1052abf10>, 'target': <bound method Server.run of <uvicorn.server.Server 
                             object...                                                                                                                                                      
                                 |    |        |    |        -> <SpawnProcess name='SpawnProcess-1' parent=86862 started>                                                                   
                                 |    |        |    -> ()                                                                                                                                   
                                 |    |        -> <SpawnProcess name='SpawnProcess-1' parent=86862 started>                                                                                 
                                 |    -> <function subprocess_started at 0x106224ca0>                                                                                                       
                                 -> <SpawnProcess name='SpawnProcess-1' parent=86862 started>                                                                                               
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/uvicorn/subprocess.py", line 76, in subprocess_started                              
                                 target(sockets=sockets)                                                                                                                                    
                                 |              -> [<socket.socket fd=14, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0',
    8000)>
  ]                     
                                 -> <bound method Server.run of <uvicorn.server.Server object at 0x1052abe20>>                                                                              
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/uvicorn/server.py", line 60, in run                                                 
                                 return asyncio.run(self.serve(sockets=sockets))                                                                                                            
                                        |       |   |    |             -> [<socket.socket fd=14, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0',
    8000)>
  ]                                                                                                                                                        
                                        |       |   |    -> <function Server.serve at 0x106224040>                                                                                          
                                        |       |   -> <uvicorn.server.Server object at 0x1052abe20>                                                                                        
                                        |       -> <function run at 0x10575d480>                                                                                                            
                                        -> <module 'asyncio' from '/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio...                 
                                                                                                                                                                                            
                               File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run                     
                                 return loop.run_until_complete(main)                                                                                                                       
                                        |    |                  -> <coroutine object Server.serve at 0x1061e9850>                                                                           
                                        |    -> <method 'run_until_complete' of 'uvloop.loop.Loop' objects>                                                                                 
                                        -> <uvloop.Loop running=True closed=False debug=False>                                                                                              
                                                                                                                                                                                            
                             > File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 372, in run_asgi                    
                                 result = await app(self.scope, self.receive, self.send)                                                                                                    
                                                |   |    |      |    |        |    -> <function RequestResponseCycle.send at 0x1063e5360>                                                   
                                                |   |    |      |    |        -> <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b356740>                         
                                                |   |    |      |    -> <function RequestResponseCycle.receive at 0x1063e53f0>                                                              
                                                |   |    |      -> <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b356740>                                       
                                                |   |    -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
    }, 'http_version': '1.1', 'server': ('127.0.0.1',
    8000), 'cl...
                                                |   -> <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b356740>                                                   
                                                -> <uvicorn.middleware.proxy_headers.ProxyHeadersMiddleware object at 0x105fd89d0>                                                          
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__                          
                                 return await self.app(scope, receive, send)                                                                                                                
                                              |    |   |      |        -> <bound method RequestResponseCycle.send of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at  
                             0x10b356740>>                                                                                                                                                  
                                              |    |   |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at        
                             0x10b3567...                                                                                                                                                   
                                              |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
      }, 'http_version': '1.1', 'server': ('127.0.0.1',
      8000), 'cl...  
                                              |    -> <fastapi.applications.FastAPI object at 0x109fc3a90>                                                                                  
                                              -> <uvicorn.middleware.proxy_headers.ProxyHeadersMiddleware object at 0x105fd89d0>                                                            
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/fastapi/applications.py", line 261, in __call__                                     
                                 await super().__call__(scope, receive, send)                                                                                                               
                                                        |      |        -> <bound method RequestResponseCycle.send of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 
                             0x10b356740>>                                                                                                                                                  
                                                        |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at       
                             0x10b3567...                                                                                                                                                   
                                                        -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
        }, 'http_version': '1.1', 'server': ('127.0.0.1',
        8000), 'cl... 
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/applications.py", line 119, in __call__                                   
                                 await self.middleware_stack(scope, receive, send)                                                                                                          
                                       |    |                |      |        -> <bound method RequestResponseCycle.send of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle      
                             object at 0x10b356740>>                                                                                                                                        
                                       |    |                |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at  
                             0x10b3567...                                                                                                                                                   
                                       |    |                -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
          }, 'http_version': '1.1', 'server': ('127.0.0.1',
          8000),   
                             'cl...                                                                                                                                                         
                                       |    -> <starlette.middleware.errors.ServerErrorMiddleware object at 0x10af37d30>                                                                    
                                       -> <fastapi.applications.FastAPI object at 0x109fc3a90>                                                                                              
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 181, in __call__                              
                                 raise exc                                                                                                                                                  
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 159, in __call__                              
                                 await self.app(scope, receive, _send)                                                                                                                      
                                       |    |   |      |        -> <function ServerErrorMiddleware.__call__.<locals>._send at 0x10b32f370>                                                  
                                       |    |   |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b3567...  
                                       |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
            }, 'http_version': '1.1', 'server': ('127.0.0.1',
            8000), 'cl...         
                                       |    -> <Companylib.console.server.core.middleware.CompressionMiddleware object at 0x10af37d00>                                                       
                                       -> <starlette.middleware.errors.ServerErrorMiddleware object at 0x10af37d30>                                                                         
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/src/backend/Companylib/console/server/core/middleware.py", line 164, in __call__                                        
                                 await responder(scope, receive, send)                                                                                                                      
                                       |         |      |        -> <function ServerErrorMiddleware.__call__.<locals>._send at 0x10b32f370>                                                 
                                       |         |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b3567... 
                                       |         -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
              }, 'http_version': '1.1', 'server': ('127.0.0.1',
              8000), 'cl...        
                                       -> <starlette.middleware.gzip.GZipResponder object at 0x10b355ae0>                                                                                   
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 43, in __call__                                 
                                 await self.app(scope, receive, self.send_with_gzip)                                                                                                        
                                       |    |   |      |        |    -> <function GZipResponder.send_with_gzip at 0x10af56e60>                                                              
                                       |    |   |      |        -> <starlette.middleware.gzip.GZipResponder object at 0x10b355ae0>                                                          
                                       |    |   |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b3567...  
                                       |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
                }, 'http_version': '1.1', 'server': ('127.0.0.1',
                8000), 'cl...         
                                       |    -> <Companylib.console.server.core.middleware.DatabaseSessionMiddleware object at 0x10af37ca0>                                                   
                                       -> <starlette.middleware.gzip.GZipResponder object at 0x10b355ae0>                                                                                   
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 66, in __call__                                 
                                 response = await self.dispatch_func(request, call_next)                                                                                                    
                                                  |    |             |        -> <function BaseHTTPMiddleware.__call__.<locals>.call_next at 0x10b32f910>                                   
                                                  |    |             -> <starlette.requests.Request object at 0x10b3550c0>                                                                  
                                                  |    -> <bound method DatabaseSessionMiddleware.dispatch of <Companylib.console.server.core.middleware.DatabaseSessionMiddleware objec...  
                                                  -> <Companylib.console.server.core.middleware.DatabaseSessionMiddleware object at 0x10af37ca0>                                             
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/src/backend/Companylib/console/server/core/middleware.py", line 90, in dispatch                                         
                                 response = await call_next(request)                                                                                                                        
                                                  |         -> <starlette.requests.Request object at 0x10b3550c0>                                                                           
                                                  -> <function BaseHTTPMiddleware.__call__.<locals>.call_next at 0x10b32f910>                                                               
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 44, in call_next                                
                                 raise app_exc                                                                                                                                              
                                       -> TypeError("'<' not supported between instances of 'NoneType' and 'int'")                                                                          
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 34, in coro                                     
                                 await self.app(scope, request.receive, send_stream.send)                                                                                                   
                                       |    |   |      |       |        |           -> <function MemoryObjectSendStream.send at 0x106380550>                                                
                                       |    |   |      |       |        -> MemoryObjectSendStream(_state=MemoryObjectStreamState(max_buffer_size=0, buffer=deque([]), open_send_channels=0, 
                             open_receive...                                                                                                                                                
                                       |    |   |      |       -> <property object at 0x108114310>                                                                                          
                                       |    |   |      -> <starlette.requests.Request object at 0x10b3550c0>                                                                                
                                       |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
                  }, 'http_version': '1.1', 'server': ('127.0.0.1',
                  8000), 'cl...         
                                       |    -> <starlette.middleware.cors.CORSMiddleware object at 0x10af37c70>                                                                             
                                       -> <Companylib.console.server.core.middleware.DatabaseSessionMiddleware object at 0x10af37ca0>                                                        
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/cors.py", line 84, in __call__                                 
                                 await self.app(scope, receive, send)                                                                                                                       
                                       |    |   |      |        -> <bound method MemoryObjectSendStream.send of MemoryObjectSendStream(_state=MemoryObjectStreamState(max_buffer_size=0,    
                             buffer=...                                                                                                                                                     
                                       |    |   |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b3567...  
                                       |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
                    }, 'http_version': '1.1', 'server': ('127.0.0.1',
                    8000), 'cl...         
                                       |    -> <Companylib.console.server.core.middleware.SecurityHeaderMiddleware object at 0x10af37be0>                                                    
                                       -> <starlette.middleware.cors.CORSMiddleware object at 0x10af37c70>                                                                                  
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 66, in __call__                                 
                                 response = await self.dispatch_func(request, call_next)                                                                                                    
                                                  |    |             |        -> <function BaseHTTPMiddleware.__call__.<locals>.call_next at 0x10b32f640>                                   
                                                  |    |             -> <starlette.requests.Request object at 0x10b356f50>                                                                  
                                                  |    -> <bound method SecurityHeaderMiddleware.dispatch of <Companylib.console.server.core.middleware.SecurityHeaderMiddleware object ...  
                                                  -> <Companylib.console.server.core.middleware.SecurityHeaderMiddleware object at 0x10af37be0>                                              
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/src/backend/Companylib/console/server/core/middleware.py", line 33, in dispatch                                         
                                 response = await call_next(request)                                                                                                                        
                                                  |         -> <starlette.requests.Request object at 0x10b356f50>                                                                           
                                                  -> <function BaseHTTPMiddleware.__call__.<locals>.call_next at 0x10b32f640>                                                               
                                                                                                                                                                                            
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 44, in call_next                                
                                 raise app_exc                                                                                                                                              
                                       -> TypeError("'<' not supported between instances of 'NoneType' and 'int'")                                                                          
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 34, in coro                                     
                                 await self.app(scope, request.receive, send_stream.send)                                                                                                   
                                       |    |   |      |       |        |           -> <function MemoryObjectSendStream.send at 0x106380550>                                                
                                       |    |   |      |       |        -> MemoryObjectSendStream(_state=MemoryObjectStreamState(max_buffer_size=0, buffer=deque([]), open_send_channels=0, 
                             open_receive...                                                                                                                                                
                                       |    |   |      |       -> <property object at 0x108114310>                                                                                          
                                       |    |   |      -> <starlette.requests.Request object at 0x10b356f50>                                                                                
                                       |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
                      }, 'http_version': '1.1', 'server': ('127.0.0.1',
                      8000), 'cl...         
                                       |    -> <starlette.exceptions.ExceptionMiddleware object at 0x10af37b80>                                                                             
                                       -> <Companylib.console.server.core.middleware.SecurityHeaderMiddleware object at 0x10af37be0>                                                         
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/exceptions.py", line 90, in __call__                                      
                                 raise exc                                                                                                                                                  
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/exceptions.py", line 79, in __call__                                      
                                 await self.app(scope, receive, sender)                                                                                                                     
                                       |    |   |      |        -> <function ExceptionMiddleware.__call__.<locals>.sender at 0x10b32fe20>                                                   
                                       |    |   |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b3567...  
                                       |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
                        }, 'http_version': '1.1', 'server': ('127.0.0.1',
                        8000), 'cl...         
                                       |    -> <fastapi.middleware.asyncexitstack.AsyncExitStackMiddleware object at 0x10af379a0>                                                           
                                       -> <starlette.exceptions.ExceptionMiddleware object at 0x10af37b80>                                                                                  
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__                         
                                 raise e                                                                                                                                                    
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__                         
                                 await self.app(scope, receive, send)                                                                                                                       
                                       |    |   |      |        -> <function ExceptionMiddleware.__call__.<locals>.sender at 0x10b32fe20>                                                   
                                       |    |   |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b3567...  
                                       |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
                          }, 'http_version': '1.1', 'server': ('127.0.0.1',
                          8000), 'cl...         
                                       |    -> <fastapi.routing.APIRouter object at 0x109fc3250>                                                                                            
                                       -> <fastapi.middleware.asyncexitstack.AsyncExitStackMiddleware object at 0x10af379a0>                                                                
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/routing.py", line 666, in __call__                                        
                                 await route.handle(scope, receive, send)                                                                                                                   
                                       |     |      |      |        -> <function ExceptionMiddleware.__call__.<locals>.sender at 0x10b32fe20>                                               
                                       |     |      |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at           
                             0x10b3567...                                                                                                                                                   
                                       |     |      -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
                            }, 'http_version': '1.1', 'server': ('127.0.0.1',
                            8000), 'cl...     
                                       |     -> <function Route.handle at 0x10813fac0>                                                                                                      
                                       -> <fastapi.routing.APIRoute object at 0x10b082da0>                                                                                                  
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/routing.py", line 266, in handle                                          
                                 await self.app(scope, receive, send)                                                                                                                       
                                       |    |   |      |        -> <function ExceptionMiddleware.__call__.<locals>.sender at 0x10b32fe20>                                                   
                                       |    |   |      -> <bound method RequestResponseCycle.receive of <uvicorn.protocols.http.httptools_impl.RequestResponseCycle object at 0x10b3567...  
                                       |    |   -> {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.3'
                              }, 'http_version': '1.1', 'server': ('127.0.0.1',
                              8000), 'cl...         
                                       |    -> <function request_response.<locals>.app at 0x10b07caf0>                                                                                      
                                       -> <fastapi.routing.APIRoute object at 0x10b082da0>                                                                                                  
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/routing.py", line 65, in app                                              
                                 response = await func(request)                                                                                                                             
                                                  |    -> <starlette.requests.Request object at 0x10b356b90>                                                                                
                                                  -> <function get_request_handler.<locals>.app at 0x10b07ca60>                                                                             
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/fastapi/routing.py", line 217, in app                                               
                                 solved_result = await solve_dependencies(                                                                                                                  
                                                       -> <function solve_dependencies at 0x10813df30>                                                                                      
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/fastapi/dependencies/utils.py", line 465, in solve_dependencies                     
                                 response = response or Response(                                                                                                                           
                                            |           -> <class 'starlette.responses.Response'>                                                                                           
                                            -> None                                                                                                                                         
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/responses.py", line 50, in __init__                                       
                                 self.init_headers(headers)                                                                                                                                 
                                 |    |            -> None                                                                                                                                  
                                 |    -> <function Response.init_headers at 0x1081193f0>                                                                                                    
                                 -> <starlette.responses.Response object at 0x10b356c80>                                                                                                    
                               File "/Users/hp/MyCode/Company/console/.venv/lib/python3.10/site-packages/starlette/responses.py", line 77, in init_headers                                   
                                 and not (self.status_code < 200 or self.status_code in (204,
                              304))                                                                                         
                                          |    |                    |    -> None                                                                                                            
                                          |    |                    -> <starlette.responses.Response object at 0x10b356c80>                                                                 
                                          |    -> None                                                                                                                                      
                                          -> <starlette.responses.Response object at 0x10b356c80>                                                                                           
                                                                                                                                                                                            
                             TypeError: '<' not supported between instances of 'NoneType' and 'int'                                                                                         
                                                                                                                                                                                            
                             TypeError: '<' not supported between instances of 'NoneType' and 'int'                                                                                         
[
                                04/08/22 10: 55: 02
                              ] INFO     127.0.0.1: 57515 - "GET /favicon.ico HTTP/1.1" 404
```

**Comment by hiranp** [comment]
Also, I going to update my pull request, found another corner case. I am not handling.

**Comment by Kludex** [comment]
How can I reproduce the issue here?

**Comment by hiranp** [comment]
This issues seems to occur on a serving static files and this case from the linked issues from FastApi [4050]. However, I need to find time to translate my specific case into a smaller test case. 
```

from typing import Awaitable, Callable

from fastapi import FastAPI
from fastapi.middleware.gzip import GZipMiddleware
from starlette.requests import Request

app = FastAPI()


@app.middleware("http")
async def noop_middleware(request: Request, call_next: Callable[[Request], Awaitable]):
    return await call_next(request)


@app.get("/", status_code=304)
def youve_seen_this_before():
    pass


app.add_middleware(GZipMiddleware)
```

**Comment by Kludex** [comment]
I'm able to reproduce this with:

```python
from typing import Awaitable, Callable

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.middleware.gzip import GZipMiddleware
from starlette.routing import Route
from starlette.middleware import Middleware
from starlette.applications import Starlette
from starlette.responses import Response
from starlette.requests import Request

class NoopMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: Callable[[Request], Awaitable]):
        return await call_next(request)

def endpoint(request: Request):
    return Response(status_code=304)


app2 = Starlette(
    routes=[Route("/", endpoint=endpoint)],
    middleware=[Middleware(GZipMiddleware), Middleware(NoopMiddleware)]
)
```
This is more as a self note. I'll continue to check this later on.

**Comment by abersheeran** [comment]
This PR did solve the problem, but that `# noqa` made me confused. 🤔

**Comment by levrik** [comment]
These seem to be linting rules coming from https://wemake-python-stylegui.de/en/latest/
Probably @hiranp has this set up globally on his machine?

File: starlette/middleware/gzip.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
         await self.app(scope, receive, send)
 
 
-class GZipResponder:
+class GZipResponder:  # noqa: WPS230
```
Comment (Kludex) [inline starlette/middleware/gzip.py]:
```suggestion
class GZipResponder:
```

File: starlette/middleware/gzip.py [hunk 2]
```diff
@@ -35,8 +35,7 @@ def __init__(self, app: ASGIApp, minimum_size: int, compresslevel: int = 9) -> N
         self.started = False
         self.gzip_buffer = io.BytesIO()
         self.gzip_file = gzip.GzipFile(
-            mode="wb", fileobj=self.gzip_buffer, compresslevel=compresslevel
-        )
+            mode="wb", fileobj=self.gzip_buffer, compresslevel=compresslevel)
```
Comment (Kludex) [inline starlette/middleware/gzip.py]:
```suggestion
            mode="wb", fileobj=self.gzip_buffer, compresslevel=compresslevel
        )
```

File: starlette/middleware/gzip.py [hunk 3]
```diff
@@ -56,11 +55,9 @@ async def send_with_gzip(self, message: Message) -> None:
                 # Don't apply GZip to small outgoing responses.
                 await self.send(self.initial_message)
                 await self.send(message)
-            elif not more_body:
+            elif not more_body:  # noqa: WPS504
```
Comment (Kludex) [inline starlette/middleware/gzip.py]:
```suggestion
            elif not more_body:
```

File: starlette/middleware/gzip.py [hunk 4]
```diff
@@ -75,13 +72,9 @@ async def send_with_gzip(self, message: Message) -> None:
                 headers = MutableHeaders(raw=self.initial_message["headers"])
                 headers["Content-Encoding"] = "gzip"
                 headers.add_vary_header("Accept-Encoding")
-                del headers["Content-Length"]
-
-                self.gzip_file.write(body)
-                message["body"] = self.gzip_buffer.getvalue()
-                self.gzip_buffer.seek(0)
-                self.gzip_buffer.truncate()
+                del headers["Content-Length"]  # noqa: WPS420
```
Comment (Kludex) [inline starlette/middleware/gzip.py]:
```suggestion
                del headers["Content-Length"]
```

**Comment by Kludex** [comment]
This PR doesn't solve the issue for me...

I've tried with both examples above, and I still have the same issues.

**Comment by Kludex** [comment]
Thanks for the PR @hiranp !

After debugging this issue, it looks like it's more related to the `BaseHTTPMiddleware` than the `GZipMiddleware`.

I'll be closing this issue, and I'd really appreciate if you can take a look at #1609. 🙏

**Pull request closed without merging**